File: backend/aws_interaction.nim
import std/[osproc, strformat, os, json, logging]
import config
import storage_local
import database_local

proc awsArgs(cfg: Config): seq[string] =
  result = @["--region", cfg.awsRegion]
  if cfg.awsProfile.len > 0:
    result.add(@["--profile", cfg.awsProfile])

proc ensureAwsCli*(): bool =
  try:
    discard execProcess("aws", args = @["--version"], options = {poUsePath, poStdErrToStdOut})
    result = true
  except OSError:
    result = false

proc uploadToS3*(cfg: Config, key: string, data: string): string =
  if cfg.mode == pmAws:
    if not ensureAwsCli():
      raise newException(OSError, "AWS CLI not found. Install awscli or use local mode.")
    let filePath = getTempDir() / key
    writeFile(filePath, data)
    let args = @["s3", "cp", filePath, fmt"s3://{cfg.awsBucket}/{key}"] & awsArgs(cfg)
    let output = execProcess("aws", args = args, options = {poUsePath, poStdErrToStdOut})
    logging.info fmt"Uploaded {key} to S3 bucket {cfg.awsBucket}"
    logging.info output
    return fmt"s3://{cfg.awsBucket}/{key}"
  let stored = writeObject(cfg, key, data)
  logging.info fmt"Stored {key} locally at {stored.path}"
  stored.path

proc persistRecord*(cfg: Config, id: string, payload: JsonNode): string =
  if cfg.mode == pmAws:
    if not ensureAwsCli():
      raise newException(OSError, "AWS CLI not found. Install awscli or use local mode.")
    let args = @["dynamodb", "put-item", "--table-name", "NimTable", "--item", $payload] & awsArgs(cfg)
    let output = execProcess("aws", args = args, options = {poUsePath, poStdErrToStdOut})
    logging.info fmt"Persisted record {id} to DynamoDB"
    logging.info output
    return "aws://dynamodb/NimTable"
  let record = appendRecord(cfg, id, payload)
  logging.info fmt"Persisted record {id} locally at {record.path}"
  record.path

File: backend_py/aws_interaction.py
from __future__ import annotations

import json
import subprocess
from typing import Dict, Any

from .config import Config
from .storage import write_object
from .database import append_record


def _aws_args(config: Config) -> list[str]:
    args = ["--region", config.aws_region]
    if config.aws_profile:
        args += ["--profile", config.aws_profile]
    return args


def ensure_aws_cli() -> bool:
    try:
        subprocess.run(["aws", "--version"], check=True, capture_output=True)
        return True
    except (OSError, subprocess.CalledProcessError):
        return False


def upload_to_s3(config: Config, key: str, data: str) -> str:
    if config.mode == "aws":
        if not ensure_aws_cli():
            raise OSError("AWS CLI not found. Install awscli or use local mode.")
        tmp_path = write_object(config, key, data).path
        cmd = ["aws", "s3", "cp", str(tmp_path), f"s3://{config.aws_bucket}/{key}"]
        subprocess.run(cmd + _aws_args(config), check=True)
        return f"s3://{config.aws_bucket}/{key}"
    stored = write_object(config, key, data)
    return str(stored.path)


def persist_record(config: Config, record_id: str, payload: Dict[str, Any]) -> str:
    if config.mode == "aws":
        if not ensure_aws_cli():
            raise OSError("AWS CLI not found. Install awscli or use local mode.")
        cmd = [
            "aws",
            "dynamodb",
            "put-item",
            "--table-name",
            "NimTable",
            "--item",
            json.dumps(payload),
        ]
        subprocess.run(cmd + _aws_args(config), check=True)
        return "aws://dynamodb/NimTable"
    record = append_record(config, record_id, payload)
    return str(record.path)
